doc/source/data/transforming-data.rst

```diff
@@ -1,286 +1,227 @@
-==============
+        .. testcode::
 
-Call :meth:`Dataset.random_shuffle() <ray.data.Dataset.random_shuffle>` to
-globally shuffle the order of data records.
+            from typing import Dict
+            import numpy as np
+            import ray
 
-.. doctest::
+            items = [
+                {"image": np.zeros((32, 32, 3)), "label": i}
+                for _ in range(10) for label in range(100)
+            ]
 
-    >>> import ray
-    >>> dataset = ray.data.range(10)
-    >>> dataset.random_shuffle().take_batch()  # doctest: +SKIP
-    {'id': array([7, 0, 9, 3, 5, 1, 4, 2, 8, 6])}
+            def normalize_images(group: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:
+                group["image"] = (group["image"] - group["image"].mean()) / group["image"].std()
+                return group
 
-For reduced overhead during training ingest, use local shuffles. Read 
-:ref:`Shuffling Data <air-shuffle>` in the AIR user guide to learn more.
+            ds = (
+                ray.data.from_items(items)
+                .groupby("label")
+                .map_groups(normalize_images)
+            )
 
-.. _data-groupbys:
+    .. tab-item:: pandas
 
-------------------
-Grouped transforms
-------------------
+        .. testcode::
 
-Ray Data supports grouping data by column and applying aggregations to each group. This is supported via the :meth:`ds.groupby() <ray.data.Dataset.groupby>` call.
+            import pandas as pd
+            import ray
 
-Aggregations
-============
+            def normalize_features(group: pd.DataFrame) -> pd.DataFrame:
+                target = group.drop("target")
+                group = (group - group.min()) / group.std()
+                group["target"] = target
+                return group
 
-Aggregations can be performed per group:
+            ds = (
+                ray.data.read_csv("s3://anonymous@air-example-data/iris.csv")
+                .groupby("target")
+                .map_groups(normalize_features)
+            )
 
-.. code-block:: python
+Shuffling rows
+==============
 
-    ds = ray.data.from_items([
-        {"A": x % 3, "B": 2 * x, "C": 3 * x}
-        for x in range(10)
-    ])
+.. testcode::
 
-    # Group by the A column and calculate the per-group mean for B and C columns.
-    ds.groupby("A").mean(["B", "C"]).to_pandas()
-    # ->
-    #    A  mean(B)  mean(C)
-    # 0  0      9.0     13.5
-    # 1  1      8.0     12.0
-    # 2  2     10.0     15.0
+    import ray
 
-Aggregations can also be applied globally:
+    ds = (
+        ray.data.read_images("example://image-datasets/simple")
+        .random_shuffle()
+    )
 
-.. code-block:: python
+.. tip::
 
-    from ray.data.aggregate import Mean, Std
+    :meth:`~ray.data.Dataset.random_shuffle` is slow. For better performance, try
+    :ref:`iterating-over-batches-with-shuffling`.
 
-    # Global mean on B and C columns.
-    ds.mean(["B", "C"])
-    # -> {'mean(B)': 9.0, 'mean(C)': 13.5}
+Creating rows
```

Comment:
I think creating rows is a bit of weird term. Both `map` and `flat_map` should be in "Transforming Rows" section. And then differentiate between the two by saying 1:1 is map, 1:N is flat_map